src/librustc_trans/back/write.rs

```diff
@@ -1109,164 +1201,230 @@ fn execute_work<'a>(sess: &'a Session,
-                                               &diag.msg,
-                                               &code,
-                                               diag.lvl);
+            // Relinquish accidentally acquired extra tokens
+            tokens.truncate(running);
+
+            match coordinator_receive.recv().unwrap() {
+                // Save the token locally and the next turn of the loop will use
+                // this to spawn a new unit of work, or it may get dropped
+                // immediately if we have no more work to spawn.
+                Message::Token(token) => {
+                    if let Ok(token) = token {
+                        tokens.push(token);
+                    } else {
+                        shared_emitter.fatal("failed to acquire jobserver token");
+                        // Exit the coordinator thread
+                        panic!()
                     }
-                    None => {
-                        handler.emit(&MultiSpan::new(),
-                                     &diag.msg,
-                                     diag.lvl);
+                }
+
+                Message::TranslationDone { llvm_work_item, is_last } => {
+                    work_items.insert(0, llvm_work_item);
+
+                    if is_last {
+                        // If this is the last, don't request a token because
+                        // the trans worker thread will be free to handle this
+                        // immediately.
+                        translation_done = true;
+                    } else {
+                        helper.request_token();
                     }
+
+                    assert_eq!(trans_worker_state, TransWorkerState::Translating);
+                    trans_worker_state = TransWorkerState::Idle;
                 }
-            }
-            Message::InlineAsmError(cookie, msg) => {
-                match Mark::from_u32(cookie).expn_info() {
-                    Some(ei) => sess.span_err(ei.call_site, &msg),
-                    None     => sess.err(&msg),
+
+                // If a thread exits successfully then we drop a token associated
+                // with that worker and update our `running` count. We may later
+                // re-acquire a token to continue running more work. We may also not
+                // actually drop a token here if the worker was running with an
+                // "ephemeral token"
+                //
+                // Note that if the thread failed that means it panicked, so we
+                // abort immediately.
+                Message::Done { result: Ok(compiled_module), worker_id } => {
+                    if worker_id == TRANS_WORKER_ID {
+                        assert_eq!(trans_worker_state, TransWorkerState::LLVMing);
+                        trans_worker_state = TransWorkerState::Idle;
+                    } else {
+                        drop(tokens.pop());
```

Comment:
Yes, good idea. The `truncate` above should take care of dropping the `Token` if we actually don't need it.